### DISCUSSION TITLE
Is there a way to GDB-like state on crash all the time?

### DISCUSSION THREAD

**Discussion by kxygk** [clojure-emacs/cider#2999] [Q&A]
I didn't know about this section on github before I posted this question on reddit:
https://old.reddit.com/r/Clojure/comments/mh38k7/a_couple_gdb_features_i_miss_in_cider/

But is there some way to enable the debugger globally to always get a program state on each crash? ie. instrument everything.. (with a performance penalty)

Often I will have my GUI thread trigger some callback and then I get a crash and a stack trace. Then I need to decipher the Clojure errors and consult my astrologist to figure out what may have caused things to blow up (ex: which arg was actually Null there..?). Is there some way I could get the program to always "stop" where the crash happened and then I can inspect the program state at that point? (including local variable in the callback scope)

At the moment, after each crash I end up having to pepper things with print statements to dump arg values. Then when it crashes you look at the printout and see what was the last thing printed. Or you cook up some test or crash-triggering-example and then inspecting with the debugger. Either way, it really breaks up the workflow and I end up with the feeling "i'm doing this wrong.."

My point of reference is the default behavior under GDB where it crashes and gives you a stack but then drops you into a mini-REPL at the crash point. Where you can poke around and even run functions (though there are some limitations). In fact you can often even "fix up" the locals vars and continue the program

**Comment by bbatsov** [comment]
> But is there some way to enable the debugger globally to always get a program state on each crash? ie. instrument everything.. (with a performance penalty)

Currently that's not possible, although I recall we've discussed something like this in the past. The behaviour you describe is also consistent with how SLIME's debugger for CL works, and how Emacs's own Elisp debugger works.

We might get there some day, but it will require a lot of work and lately CIDER has been short on contributors, so things are moving relatively slowly.

**Reply by kxygk** [reply]
I had the feeling it's how ELisp works (and Cider is inspired by that). Just haven't used it heavily in a while :)

Anyway, thanks for the sanity check. I thought maybe I was missing something obvious.
